### PR TITLE
feat(ai-gemini): stream structured outputs natively

### LINKED ISSUE #570: feat(ai-gemini): native structuredOutputStream for geminiText and geminiTextInteractions
## Background

Both Gemini text adapters currently fall through to `fallbackStructuredOutputStream` in `@tanstack/ai` ([`packages/typescript/ai/src/activities/chat/index.ts:1796-1869`](https://github.com/TanStack/ai/blob/main/packages/typescript/ai/src/activities/chat/index.ts#L1796-L1869)) when called with `chat({ outputSchema, stream: true })`:

- `geminiText()` — `packages/typescript/ai-gemini/src/adapters/text.ts`
- `geminiTextInteractions()` (experimental) — `packages/typescript/ai-gemini/src/experimental/text-interactions/adapter.ts`

Neither implements the optional `structuredOutputStream` adapter method, so the core helper calls the non-streaming `structuredOutput()`, **buffers the entire JSON**, then re-emits it as a single synthetic `TEXT_MESSAGE_CONTENT` chunk followed by a `structured-output.complete` CUSTOM event.

This is "stream-compatible" but **not truly streaming** — users see one big chunk arrive at once, not incremental JSON deltas.

For comparison, the OpenAI-base-derived adapters (openai, openrouter chat-completions + responses, grok, groq) all implement `structuredOutputStream` natively — see `packages/typescript/openai-base/src/adapters/{chat-completions-text.ts:231,responses-text.ts:270}`.

### `geminiText()`

Use `client.models.generateContentStream(...)` with `responseMimeType: 'application/json'` and `responseSchema` (already used by the non-streaming `structuredOutput()` at `text.ts:181`). Yield incremental `TEXT_MESSAGE_CONTENT` chunks as JSON fragments arrive. On stream end, parse the accumulated text and emit the terminal `structured-output.complete` CUSTOM event with `{ object, raw }`.

### `geminiTextInteractions()` (experimental)

Use `client.interactions.create({ stream: true, response_mime_type: 'application/json', response_format: outputSchema })` (current shape already lives in the non-streaming `structuredOutput()` at `adapter.ts:152-154`). Reuse the existing `translateInteractionEvents` SSE translator — `content.delta` events with `delta.type === 'text'` carry the incremental JSON. Emit the terminal `structured-output.complete` CUSTOM event before `RUN_FINISHED`, matching the existing `gemini.interactionId` ordering invariant.

## Tests to add (`packages/typescript/ai-gemini/tests/`)

- `text-adapter.test.ts` — call `chat({ adapter: geminiText(...), outputSchema, stream: true })`, assert ≥2 `TEXT_MESSAGE_CONTENT` chunks (proves incremental streaming, not buffered), and that the terminal `structured-output.complete` CUSTOM event carries the parsed object.
- `text-interactions-adapter.test.ts` — same shape against `geminiTextInteractions()`; also assert `gemini.interactionId` is still emitted before `RUN_FINISHED` when structured streaming.

## E2E

The `structured-output-stream` feature already exists in `testing/e2e/src/lib/feature-support.ts`. Adding `gemini` to its support set + a fixture under `testing/e2e/fixtures/structured-output-stream/` should be enough — the existing `structured-output-stream.spec.ts` covers the `contentDeltaCount > 1` invariant via the route's `onChunk` counter.

## Out of scope

- Image / audio / video output modalities through structured streaming — those flow through `geminiImage`/`geminiSpeech` and have their own adapter surfaces.
- Tool-call interleaving inside structured-output streams — the OpenAI-base implementations also don't mix tool calls into a single structured stream; same constraint here.

## Context

Surfaced during PR review of #502 (`feat(ai-gemini): add geminiTextInteractions() adapter for stateful Interactions API`). #502 itself ships the non-streaming `structuredOutput()` only; this issue tracks the streaming variant for both Gemini adapters as a follow-up.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by Missing-Identity**
## What changed

- add native `structuredOutputStream` support to the standard Gemini text adapter
- add native structured-output streaming to the experimental Gemini Interactions adapter
- accumulate and parse streamed JSON before emitting `structured-output.complete`
- preserve interaction ID and run-finished event ordering
- add focused adapter coverage, enable Gemini in the structured-output-stream E2E matrix, and add a minor changeset

Closes #570.

## Validation

- focused Gemini adapter tests: 46 passed
- complete `@tanstack/ai-gemini` test suite: 239 passed
- `@tanstack/ai-gemini` TypeScript check: passed
- `@tanstack/ai-gemini` ESLint: no errors; seven pre-existing warnings
- `@tanstack/ai-gemini` production build: passed
- `git diff --check`: passed

## Limitations

The hosted environment could not download Playwright Chromium, so browser E2E execution was not claimed. The existing structured-output-stream E2E matrix was updated to include Gemini.

## Summary by CodeRabbit

* **New Features**
  * Added native structured-output streaming for Gemini text generation.
  * Added structured-output streaming support for the experimental Gemini Interactions API.
  * Streams now emit parsed JSON results and original text before final completion.
* **Improved Reliability**
  * Enhanced handling and reporting for empty, truncated, invalid, or provider-failing streams.
* **Tests**
  * Expanded adapter and end-to-end coverage for Gemini structured-output streaming.

### CODE REVIEW

**Comment by tombeckenham** [comment]
Thanks for the PR, @Missing-Identity! 🙌 @jherr will take a look.

**Automated pre-review checks**
- ✅ CI passing
- ✅ No merge conflicts
- ✅ Changeset present
- ✅ E2E test changes included

<sub>Automated triage — a human review follows.</sub>

**Comment by tombeckenham** [comment]
Maintainer sweep: rebased onto `main` (force-with-lease). CI approval handled if it was waiting.

**Comment by tombeckenham** [comment]
Maintainer sweep: rebased onto `main` (force-with-lease). CI approval handled if it was waiting.

**Comment by tombeckenham** [comment]
Maintainer sweep: rebased onto `main` (force-with-lease).

File: packages/ai-gemini/src/experimental/text-interactions/adapter.ts [hunk 1]
```diff
@@ -463,6 +463,148 @@ export class GeminiTextInteractionsAdapter<
...
+
+    try {
+      chatOptions.logger.request(
+        `activity=structuredOutputStream provider=gemini-text-interactions model=${this.model} messages=${chatOptions.messages.length}`,
+        { provider: this.name, model: this.model, request },
+      )
+      const stream = (await this.client.interactions.create(
+        request as GeminiInteractionsRequestBody &
+          Parameters<typeof this.client.interactions.create>[0],
+        { signal: chatOptions.abortController?.signal },
```
Comment (AlemTuzlak) [inline packages/ai-gemini/src/experimental/text-interactions/adapter.ts]: **[bug]** The new Interactions stream method aborts with `chatOptions.abortController?.signal`. `chat()` never puts `abortController` on structured-output `chatOptions`. It only forwards abort as `chatOptions.request.signal` (`packages/ai/src/activities/chat/index.ts` builds `effectiveRequest` from the caller abort controller and passes `{ request: this.effectiveRequest }` into `structuredCallOptions`). So `chat({ abortController, outputSchema, stream: true })` will not cancel the Google Interactions HTTP request. The sibling Gemini text method gets this right because `mapCommonOptionsToGemini` already maps `options.request.signal` to `config.abortSignal` (#1375). `chatStream` on this same adapter has the same abortController-only pattern, but this is new streaming code on the public structured-output path.

**Suggestion:** Pass `chatOptions.request?.signal ?? chatOptions.abortController?.signal` into `interactions.create`, matching how `chat()` actually forwards abort.

File: packages/ai-gemini/tests/gemini-adapter.test.ts [hunk 2]
```diff
@@ -1082,6 +1082,67 @@ describe('GeminiAdapter through AI', () => {
...
+              finishReason: 'STOP',
+            },
+          ],
+          usageMetadata: { totalTokenCount: 4 },
+        },
+      ]),
+    )
+
+    const events: Array<AdapterYieldChunk> = []
+    for await (const event of createTextAdapter().structuredOutputStream({
```
Comment (AlemTuzlak) [inline packages/ai-gemini/tests/gemini-adapter.test.ts]: **[suggestion]** Issue #570 asked for tests through `chat({ adapter, outputSchema, stream: true })`. Both new happy-path tests call `adapter.structuredOutputStream()` directly. That still proves request shape, two text deltas, and complete-before-finished, and `packages/ai` already tests the engine dispatch. It does not prove the public Gemini text path. The Interactions non-stream test was also moved off `chat()` onto `adapter.structuredOutput()` (needed, because `chat()` now prefers the stream method). The composition test still uses `chat({ outputSchema })` with a stream mock, so Interactions is partly covered; Gemini text is not.

**Suggestion:** Add one `chat({ adapter, outputSchema, stream: true })` test per adapter that asserts `>= 2` `TEXT_MESSAGE_CONTENT` chunks and a terminal `structured-output.complete`. Keep the direct adapter tests.

File: packages/ai-gemini/src/adapters/text.ts [hunk 3]
```diff
@@ -442,6 +442,107 @@ export class GeminiTextAdapter<
...
+          'empty-response',
+        )
+        return
+      }
+
+      let object: unknown
+      try {
+        object = JSON.parse(rawText)
+      } catch {
+        yield structuredStreamError(
```
Comment (AlemTuzlak) [inline packages/ai-gemini/src/adapters/text.ts]: **[suggestion]** The new methods add `empty-response`, `parse-error`, `truncated-stream`, and `provider-error` branches, but tests only cover the happy path. OpenAI-base and Bedrock test empty content and invalid JSON. The parse-error message also drops the JSON snippet that non-streaming `structuredOutput()` and openai-base include (`Content: ${rawText.slice(0, 200)}...`), which makes a failed stream harder to debug.

**Suggestion:** Add unit tests for empty content, invalid JSON, and a stream that never emits `RUN_FINISHED`. Include a short content snippet on `parse-error`, matching `structuredOutput()`.

File: packages/ai-gemini/src/adapters/text.ts [hunk 4]
```diff
@@ -442,6 +442,107 @@ export class GeminiTextAdapter<
     }
   }
 
+  /** Stream Gemini's native JSON response and emit the parsed object before completion. */
```
Comment (AlemTuzlak) [inline packages/ai-gemini/src/adapters/text.ts]: **[suggestion]** The new JSDoc only restates what the method does (`Stream Gemini's native JSON response and emit the parsed object before completion.`). Comments in this repo should say why, not repeat the code.

**Suggestion:** Delete the JSDoc, or replace it with a short why (for example: native stream so `chat({ outputSchema, stream: true })` does not fall through to `fallbackStructuredOutputStream`).

File: packages/ai-gemini/src/adapters/text.ts [hunk 5]
```diff
@@ -442,6 +442,107 @@ export class GeminiTextAdapter<
...
+      }
+
+      yield {
+        type: EventType.CUSTOM,
+        name: 'structured-output.complete',
+        value: { object, raw: rawText },
+        model: chatOptions.model,
+        timestamp: Date.now(),
+      }
+      yield finished
```
Comment (AlemTuzlak) [inline packages/ai-gemini/src/adapters/text.ts]: **[suggestion]** `processStreamChunks` stamps `RUN_FINISHED` when Gemini sends `finishReason`, then this method holds that chunk, emits `structured-output.complete` with a later `Date.now()`, and yields the old `RUN_FINISHED`. The adapter contract says events must be timestamped when they are emitted so timestamps follow stream order. OpenAI-base builds `RUN_FINISHED` at emit time after complete. `chat()` walks events in yield order, so this is not a functional break today, but timestamp order is inverted.

**Suggestion:** When yielding the held finished event, copy it and set `timestamp: Date.now()`, or build a fresh `RUN_FINISHED` after complete (keep `runId`, `threadId`, `usage`, `finishReason`).

**Review by AlemTuzlak — APPROVED** [review]
